update_duties/database_functions.py

```
@@ -0,0 +1,125 @@
+    try:
+        plant_table = db.Table('plant', database.metadata, autoload_with=database.engine)
+        query = db.select(plant_table)
+        response = database.connection.execute(query)
+        results = response.fetchall()
+        return [row.id for row in results]
+
+    except Exception as e:
+        raise e
+
+
+def get_active_duties(database: MSSQL_Database) -> pd.DataFrame:
+    """Retrieves pandas df of all records in db duty table without an end date."""
+    try:
+        duty_table = db.Table('duty', database.metadata, autoload_with=database.engine)
+        query = db.select(duty_table).where(duty_table.columns.end == None)
+        response = database.connection.execute(query)
+        results = response.fetchall()
+        return pd.DataFrame(results)
+
+    except Exception as e:
+        raise e
+
+
+def get_botanist_by_name(forename: str, surname: str, database: MSSQL_Database) -> pd.DataFrame:
+    """Retrieves pandas df of first record in db botanist table with specified fore- and surname."""
+    try:
+        botanist_table = db.Table('botanist', database.metadata, autoload_with=database.engine)
+        query = db.select(botanist_table).where(botanist_table.columns.firstname == forename and
+                                                botanist_table.columns.lastname == surname)
+        response = database.connection.execute(query)
+        result = response.first()
+        return list(result)
+
+    except Exception as e:
+        raise e
+
+
+def get_table_records(table_name: str, database: MSSQL_Database) -> pd.DataFrame:
+    """Retrieves all records as pandas df from specified table in db."""
+    try:
+        table = db.Table(table_name, database.metadata, autoload_with=database.engine)
+        query = db.select(table)
+        response = database.connection.execute(query)
+        results = response.fetchall()
+        return pd.DataFrame(results)
+
+    except Exception as e:
```

Do we definitely want to raise these errors? could it cause the script to stop running at any point?

---

It depends on when it's used.... this function isn't actually used in the file `update_duties.py`, it's a legacy from a small refactor I made, but with the view of eventually expanding this file to be a library of database functions, I thought it was useful enough to leave in.